### PR TITLE
TandemFoil: T_max=30 cosine annealing at 5L/256d baseline

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The current best TandemFoil run uses T_max=10 for cosine annealing. A longer cosine period (T_max=30) may allow the optimizer to explore a wider loss landscape before cooling, potentially reaching a better minimum. This mirrors the DrivAerML breakthrough where T_max=30 outperformed shorter schedules.

## Instructions
Run the following command exactly:

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 3e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --model-slices 64 \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --wandb-name tandem-5L256d-tmax30
```

The only change vs baseline is `--cosine-t-max 30` (was 10).

## Baseline
- TandemFoil val: 52.81 / test: 55.25
- Baseline W&B run: `l5kggnbg`
- PR: #2595, 67 epochs, 180-min wall-clock
- Reproduce: `cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 10 --no-use-ema --model-slices 64 --model-layers 5 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2710 because ADVISOR: Closing — 5L/256d+T_max=30 at lr=3e-4 is superseded by lr=2e-4..

**Comment by tcapelle** [comment]
STUDENT sakura:

## Results

### Best Result: Epoch 54 — val_primary/surface_pressure_mae = 64.62

| Metric | T_max=30 (this run, best val) | Baseline T_max=10 (PR #2595) | Delta |
|--------|------|---------|---------|
| **val_primary/surface_pressure_mae** | **64.62** (epoch 54) | 52.81 | +22.4% (worse) |
| val_single_in_dist | 69.65 | — | — |
| val_geom_camber_rc | 72.54 | — | — |
| val_geom_camber_cruise | 54.11 | — | — |
| val_re_rand | 62.17 | — | — |
| test_primary/surface_pressure_mae | 105.33* | 55.25 | +90.6% (worse) |
| Epochs | 65 (best at 54) | 67 | — |

*Test metrics are unreliable — evaluated on the post-explosion model at epoch 65, not the best epoch 54.

### Result: T_max=30 did NOT beat the T_max=10 baseline

T_max=30 underperformed T_max=10 on TandemFoil at 5L/256d. Best val was 64.62 vs baseline 52.81 (+22.4% worse). The run also suffered **two gradient explosions** that destroyed convergence:

1. **Epoch 40**: grad_norm spiked to 16.3 (from ~0.9), val jumped to 111.5. Recovery took ~8 epochs.
2. **Epoch 59**: catastrophic explosion, grad_norm = 328, val = 326.5. Model never recovered before timeout.

### Per-epoch best trajectory

| Epoch | val_primary | grad_norm | Notes |
|-------|------------|-----------|-------|
| 7 | 120.25 | 1.68 | |
| 9 | 96.75 | 1.45 | |
| 15 | 83.64 | 1.22 | |
| 21 | 78.76 | 1.10 | |
| 23 | 72.18 | 1.10 | |
| 32 | 68.51 | 1.04 | |
| 35 | 65.48 | 0.96 | |
| 40 | 111.46 | **16.3** | Gradient spike #1 |
| 43 | 71.58 | 0.99 | Recovered after ~8 epochs |
| **54** | **64.62** | 0.79 | **Best overall** |
| 59 | 326.54 | **328.0** | Gradient spike #2 — catastrophic |
| 65 | 108.16 | 1.48 | End of run, still broken |

### Analysis

- **T_max=30 is too long for TandemFoil at 5L/256d.** The slower LR decay allows gradients to remain large for longer, creating instability windows. T_max=10 reaches its minimum faster, providing more stable training.
- **Without gradient clipping, the Lion optimizer at 5L/256d is susceptible to gradient explosions.** The baseline (T_max=10) likely avoids this by having shorter cycles, so the optimizer stabilizes more frequently.
- The val trajectory before the first spike was competitive (~65 at epoch 35), but the instability events prevented the model from reaching its potential.

### Experiment details

- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 30 --no-use-ema --model-slices 64 --model-layers 5 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --agent sakura --wandb-name "sakura/tandem-5L256d-tmax30"`
- **W&B run:** tvv43g30
- **Epochs:** 65 (180-min budget)
- **GPU:** Single GPU (CUDA_VISIBLE_DEVICES=0)

### Suggested follow-ups

1. **Add gradient clipping** (e.g., `--grad-clip 1.0`) with T_max=30 to prevent the instability events. The 5L/256d model may benefit from T_max=30 if gradient explosions are prevented.
2. **T_max=10 remains the better schedule for TandemFoil** — it provides more stable training through rapid cosine cycles.
3. **T_max=15 or T_max=20** could be a middle ground — longer than 10 for broader exploration, but shorter than 30 to avoid instability windows.
